### PR TITLE
net: lwm2m: support client-initiated De-register

### PR DESCRIPTION
**Pull request by junqingzou**
Add new RD Client API of lwm2m_rd_client_stop() for this
Fix issues of de-register and event report in RD Client

### CODE REVIEW

File: subsys/net/lib/lwm2m/lwm2m_rd_client.c [hunk 1]
```diff
@@ -349,12 +349,12 @@ static int do_deregister_reply_cb(const struct coap_packet *response,
 	u8_t code;
 
 	code = coap_header_get_code(response);
-	LOG_DBG("Deregister callback (code:%u.%u)",
+	LOG_INF("Deregister callback (code:%u.%u)",
```
Comment (jukkar) [inline subsys/net/lib/lwm2m/lwm2m_rd_client.c]: IMHO these should be LOG_DBG() as typically info level prints just too much data.

File: subsys/net/lib/lwm2m/lwm2m_rd_client.c [hunk 2]
```diff
@@ -349,12 +349,12 @@ static int do_deregister_reply_cb(const struct coap_packet *response,
...
 		COAP_RESPONSE_CODE_CLASS(code),
 		COAP_RESPONSE_CODE_DETAIL(code));
 
 	if (code == COAP_RESPONSE_CODE_DELETED) {
-		LOG_DBG("Deregistration success");
+		LOG_INF("Deregistration success");
```
Comment (jukkar) [inline subsys/net/lib/lwm2m/lwm2m_rd_client.c]: same here

Comment (junqingzou) [reply subsys/net/lib/lwm2m/lwm2m_rd_client.c]: Thanks for the comment. I'm thinking that it'll be good to have de-register logging at the same level as register (as in "do_registration_reply_cb()"). App developer does not want debug prints but would be nice to know the de-register result.

Comment (jukkar) [reply subsys/net/lib/lwm2m/lwm2m_rd_client.c]: Sure, if LOG_INF() works here and does not print too much data, then it is ok. It is just that usually the info level is printed too much which is quite annoying.

**Review by mike-scott — CHANGES_REQUESTED** [review]
I think if we move the `lwm2m_engine_context_close` call and fix the minor space NIT we're in a good place for this PR.

File: subsys/net/lib/lwm2m/lwm2m_rd_client.c [hunk 3]
```diff
@@ -844,6 +848,7 @@ static void lwm2m_rd_client_service(struct k_work *work)
 			break;
 
 		case ENGINE_DEREGISTERED:
+			lwm2m_engine_context_close(client.ctx);
```
[the remark this answers is not in the document]
Comment (junqingzou) [reply subsys/net/lib/lwm2m/lwm2m_rd_client.c]: @mike-scott thanks for the comments, I updated the code accordingly.
>move it to do_deregister_reply_cb() before the line below
Sorry I moved it to "after" the line, did not read the comment carefully enough

**Review by mike-scott — APPROVED** [review]
Thanks for the update @junqingzou: LGTM

**Review by jukkar — APPROVED** [review]

**Review by pfalcon — COMMENTED** [review]
Is this correct: `RD (De-Registration)`? (Mind the order of letters in abbreviation.)

**Comment by junqingzou** [comment]
@pfalcon Thanks, updated to use the "De-register" term as in the spec.

**Review by rlubos — APPROVED** [review]
LGTM

**Pull request merged**